src/register_allocator.cc

```diff
@@ -0,0 +1,107 @@
+#include "register_allocator.h"
+
+#include <cassert>
+
+RegisterAllocator::RegisterAllocator(uint16_t num_physical_registers)
+{
+  for (uint16_t i = 0; i < num_physical_registers; ++i) {
+    free_registers.push(static_cast<PHYSICAL_REGISTER_ID>(i));
+  }
+  physical_register_file = std::vector<physical_register>(num_physical_registers, {0, 0, false, false});
+  frontend_RAT.fill(-1); //default value for no mapping
+  backend_RAT.fill(-1);
+}
+
+PHYSICAL_REGISTER_ID RegisterAllocator::rename_dest_register(int16_t reg, ooo_model_instr &instr)
+{
+  assert(free_registers.size() > 0);
+
+  PHYSICAL_REGISTER_ID phys_reg = free_registers.front();
+  free_registers.pop();
+  frontend_RAT[reg] = phys_reg;
+  physical_register_file[phys_reg] = {(uint16_t)reg, instr.instr_id, false, true}; //arch_reg_index, valid, busy
+
+  return phys_reg;
+}
+
+PHYSICAL_REGISTER_ID RegisterAllocator::rename_src_register(int16_t reg)
+{
+  PHYSICAL_REGISTER_ID phys = frontend_RAT[reg];
+
+  if (phys < 0) {
+    // allocate the register if it hasn't yet been mapped
+    // (common due to the traces being slices in the middle of a program)
+    phys = free_registers.front();
+    free_registers.pop();
+    frontend_RAT[reg] = phys;
+    physical_register_file[phys] = {(uint16_t)reg, 0, true, true}; //arch_reg_index, producing_inst_id, valid, busy
+  }
+
+  return phys;
+}
+
+void RegisterAllocator::complete_dest_register(PHYSICAL_REGISTER_ID physreg)
+{
+  // mark the physical register as valid
+  physical_register_file[physreg].valid = true;
```

Comment:
```suggestion
  physical_register_file.at(physreg).valid = true;
```

Given that the index type is signed, using the checked accessor here and throughout is probably wise.